File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/auth/CachingApiKeyAuthorizer.java
package uk.nhs.nhsx.core.auth;

import com.github.benmanes.caffeine.cache.Cache;
import com.github.benmanes.caffeine.cache.Caffeine;
import uk.nhs.nhsx.core.exceptions.Defect;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.Duration;
import java.util.Base64;
import java.util.Optional;

import static java.nio.charset.StandardCharsets.UTF_8;

public class CachingApiKeyAuthorizer implements ApiKeyAuthorizer {

    private final Base64.Encoder encoder = Base64.getEncoder();

    private final Cache<String, Boolean> cache = Caffeine.newBuilder()
        .expireAfterWrite(Duration.ofMinutes(1))
        .build();

    private final ApiKeyAuthorizer delegate;

    public CachingApiKeyAuthorizer(ApiKeyAuthorizer delegate) {
        this.delegate = delegate;
    }

    @Override
    public boolean authorize(ApiKey key) {
        try {
            return authorizeInternal(key);
        } catch (Exception e) {
            throw new RuntimeException("Unable to verify key", e);
        }
    }

    private Boolean authorizeInternal(ApiKey key) {
        return Optional.ofNullable(cache.get(hashOf(key), k -> delegate.authorize(key))).orElse(false);
    }

    private String hashOf(ApiKey key) {
        MessageDigest digest = newDigest();
        digest.update(bytesFor(key.getKeyName()));
        digest.update(bytesFor(key.getKeyValue()));
        return encoder.encodeToString(digest.digest());
    }

    private MessageDigest newDigest() {
        try {
            return MessageDigest.getInstance("SHA-256");
        } catch (NoSuchAlgorithmException e) {
            throw new Defect("Unable to get message digest", e);
        }
    }

    private byte[] bytesFor(String keyName) {
        return keyName.getBytes(UTF_8);
    }
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/auth/CachingApiKeyAuthorizer.kt
package uk.nhs.nhsx.core.auth

import com.github.benmanes.caffeine.cache.Caffeine
import uk.nhs.nhsx.core.exceptions.Defect
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.time.Duration
import java.util.Base64
import java.util.Optional

class CachingApiKeyAuthorizer(private val delegate: ApiKeyAuthorizer) : ApiKeyAuthorizer {
    private val encoder = Base64.getEncoder()
    private val cache = Caffeine.newBuilder()
        .expireAfterWrite(Duration.ofMinutes(1))
        .build<String, Boolean>()

    override fun authorize(key: ApiKey) = try {
        authorizeInternal(key)
    } catch (e: Exception) {
        throw RuntimeException("Unable to verify key", e)
    }

    private fun authorizeInternal(key: ApiKey): Boolean =
        Optional.ofNullable(cache[hashOf(key), { delegate.authorize(key) }]).orElse(false)

    private fun hashOf(key: ApiKey) = encoder.encodeToString(newDigest().apply {
        update(bytesFor(key.keyName))
        update(bytesFor(key.keyValue))
    }.digest())

    private fun newDigest() = try {
        MessageDigest.getInstance("SHA-256")
    } catch (e: NoSuchAlgorithmException) {
        throw Defect("Unable to get message digest", e)
    }

    private fun bytesFor(keyName: String) = keyName.toByteArray()
}
